hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/staging/StagingCommitter.java

```diff
@@ -0,0 +1,908 @@
+
+    // explicitly choose commit algorithm
+    initFileOutputCommitterOptions(context);
+    commitsDirectory = Paths.getMultipartUploadCommitsDirectory(conf, uuid);
+    FileSystem stagingFS = commitsDirectory.getFileSystem(conf);
+    Path qualified = stagingFS.makeQualified(commitsDirectory);
+    if (stagingFS instanceof S3AFileSystem) {
+      // currently refuse to work with S3a for the working FS; you need
+      // a consistent FS. This isn't entirely true with s3guard and
+      // alternative S3 endpoints, but doing it now stops
+      // accidental use of S3
+      throw new PathIOException(qualified.toUri().toString(),
+          "Directory for intermediate work cannot be on S3");
+    }
+    return new FileOutputCommitter(qualified, context);
+  }
+
+  /**
+   * Init the context config with everything needed for the file output
+   * committer. In particular, this code currently only works with
+   * commit algorithm 1.
+   * @param context context to configure.
+   */
+  protected void initFileOutputCommitterOptions(JobContext context) {
+    context.getConfiguration()
+        .setInt(FileOutputCommitter.FILEOUTPUTCOMMITTER_ALGORITHM_VERSION, 1);
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("StagingCommitter{");
+    sb.append(super.toString());
+    sb.append(", finalOutputPath=").append(finalOutputPath);
+    sb.append(", conflictResolution=").append(conflictResolution);
+    if (wrappedCommitter != null) {
+      sb.append(", wrappedCommitter=").append(wrappedCommitter);
+    }
+    sb.append('}');
+    return sb.toString();
+  }
+
+  /**
+   * Get the UUID of an upload; may be the job ID.
+   * @param conf job/task configuration
+   * @param jobId Job ID
+   * @return an ID for use in paths.
+   */
+  public static String getUploadUUID(Configuration conf, String jobId) {
+    return conf.getTrimmed(FS_S3A_COMMITTER_STAGING_UUID,
+        conf.get(SPARK_WRITE_UUID,
+            conf.getTrimmed(SPARK_APP_ID, jobId)));
+  }
+
+  /**
+   * Get the UUID of an upload; may be the job ID.
+   * @param conf job/task configuration
+   * @param jobId Job ID
+   * @return an ID for use in paths.
+   */
+  public static String getUploadUUID(Configuration conf, JobID jobId) {
+    return getUploadUUID(conf, jobId.toString());
+  }
+
+  /**
+   * Get the work path for a task.
+   * @param context job/task complex
+   * @param uuid UUID
+   * @return a path or null if the context is not of a task
+   * @throws IOException failure to build the path
+   */
+  private static Path buildWorkPath(JobContext context, String uuid)
+      throws IOException {
+    if (context instanceof TaskAttemptContext) {
+      return taskAttemptWorkingPath((TaskAttemptContext) context, uuid);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * The staging committers do not require "magic" commit support.
+   * @return false
+   */
+  @Override
+  protected boolean isMagicFileSystemRequired() {
```

Comment:
Is it possible to get rid of this? I never remember what "magic" refers to. I'd recommend either a better name or removing it from the API.

Reply:
I was trying really hard to think up a good name here. I will change it in this committer but use it elsewhere as "supportsPendingWritesToSpecialPaths" is just too nonintuitive